src/main/java/com/derongan/minecraft/mineinabyss/geary/AbyssLocationSystem.kt

```diff
@@ -0,0 +1,64 @@
+package com.derongan.minecraft.mineinabyss.geary
+
+import com.derongan.minecraft.deeperworld.world.section.Section
+import com.derongan.minecraft.mineinabyss.AbyssContext
+import com.derongan.minecraft.mineinabyss.world.Layer
+import com.mineinabyss.geary.ecs.Family
+import com.mineinabyss.geary.ecs.component.components.control.Activated
+import com.mineinabyss.geary.ecs.entity.GearyEntity
+import com.mineinabyss.geary.ecs.system.IteratingSystem
+import org.bukkit.ChatColor
+import org.bukkit.Location
+
+class AbyssLocationSystem(private val context: AbyssContext) : IteratingSystem(Family.builder().setAll(setOf(Activated::class.java, DepthMeter::class.java)).build()) {
+
+    override fun update(gearyEntity: GearyEntity?) {
+        val entity = gearyEntity!!
+
+        entity.holdingPlayer.ifPresent { player ->
+            val section: Section? = context.realWorldManager.getSectionFor(player.location)
+            val layer: Layer? = section?.let { context.worldManager.getLayerForSection(it) }
+
+            if (layer == null) {
+                player.sendMessage("${ChatColor.ITALIC}${ChatColor.DARK_AQUA}The compass wiggles slightly but does not otherwise respond.")
+            } else {
+                val accuracy = entity.getComponent(DepthMeter::class.java).accuracy
+                val depth = accuracy * (getDepth(layer, section, player.location) / accuracy)
+                player.sendMessage("${ChatColor.ITALIC}${ChatColor.DARK_AQUA}The compass spins. You are suddenly aware that you are about ${ChatColor.AQUA}${pluralizeMeters(depth)}${ChatColor.DARK_AQUA} deep in ${ChatColor.AQUA}${layer.name}.")
+            }
+        }
+
+        entity.removeComponent(Activated::class.java)
+    }
+
+    // TODO memoize total depth of each layer
+    private fun getDepth(layer: Layer, section: Section, location: Location): Int {
+        var totalDepth = 0
+        var currentSectionTop = 0
+        val numSections = layer.sections!!.size
+        layer.sections?.forEachIndexed { index, s ->
+            run {
```

Comment:
Does the run block do anything here?

Reply:
side effect sets currentSectionTop and totalDepth